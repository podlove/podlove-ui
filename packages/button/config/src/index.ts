import { compose, prop, propOr } from 'ramda';
import type {
  PodloveSubscribeButtonClient,
  PodloveSubscribeButtonConfig,
  PodloveTheme,
  PodloveRuntime
} from '@podlove/types';

export const feed = prop('feed') as (input: PodloveSubscribeButtonConfig) => string;
export const clients = propOr([], 'clients') as (
  input: PodloveSubscribeButtonConfig
) => PodloveSubscribeButtonClient[];
export const theme = propOr({}, 'theme') as (
  input: PodloveSubscribeButtonConfig
) => PodloveTheme;

export const runtime = propOr({}, 'runtime') as (
  input: PodloveSubscribeButtonConfig
) => PodloveRuntime;

export const language = compose(prop('language'), runtime) as (
  input: PodloveSubscribeButtonConfig
) => string;
