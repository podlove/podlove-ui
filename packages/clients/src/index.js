import antennaPod from './antenna-pod'
import applePodcasts from './apple-podcasts'
import beyondPod from './beyond-pod'
import castro from './castro'
import clementine from './clementine'
import deezer from './deezer'
import downcast from './downcast'
import googlePodcasts from './google-podcasts'
import gpodder from './gpodder'
import iCatcher from './i-catcher'
import instacast from './instacast'
import overcast from './overcast'
import playerFm from './player-fm'
import pocketCasts from './pocket-casts'
import podcastAddict from './podcast-addict'
import podcastRepublic from './podcast-republic'
import podcat from './podcat'
import podscout from './podscout'
import procast from './procast'
import rss from './rss'
import rssRadio from './rss-radio'
import soundcloud from './soundcloud'
import spotify from './spotify'
import stitcher from './stitcher'
import youtube from './youtube'
import castbox from './castbox'

const CLIENTS = {
  'antenna-pod': antennaPod,
  'apple-podcasts': applePodcasts,
  'beyond-pod': beyondPod,
  castro: castro,
  clementine: clementine,
  deezer: deezer,
  downcast: downcast,
  'google-podcasts': googlePodcasts,
  gpodder: gpodder,
  'i-catcher': iCatcher,
  instacast: instacast,
  overcast: overcast,
  'player-fm': playerFm,
  'pocket-casts': pocketCasts,
  'podcast-addict': podcastAddict,
  'podcast-republic': podcastRepublic,
  podcat: podcat,
  podscout: podscout,
  procast: procast,
  rss: rss,
  'rss-radio': rssRadio,
  soundcloud: soundcloud,
  spotify: spotify,
  stitcher: stitcher,
  youtube: youtube,
  castbox: castbox
}

const search =
  ({ id = null, platform = null, title = null, type = null } = {}) =>
  (client) => {
    const platformMatch = platform ? [].concat(platform).includes(client.platform) : true
    const typeMatch = type ? [].concat(type).includes(client.type) : true
    const titleMatch = title ? client.title.toUpperCase().includes(title.toUpperCase()) : true
    const idMatch = id ? [].concat(id).some((item) => item === client.id) : true

    return idMatch && platformMatch && titleMatch && typeMatch
  }

export default ({ id, platform, title, type } = {}) =>
  Object.keys(CLIENTS)
    .reduce((result, id) => [...result, ...CLIENTS[id].map((item) => ({ id, ...item }))], [])
    .filter(search({ id, type, title, platform }))
