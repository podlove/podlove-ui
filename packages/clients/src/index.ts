import { PodcastClient, PodcastClientType, PodcastPlatform } from './types.js';

import antennaPod from './antenna-pod/index.js';
import applePodcasts from './apple-podcasts/index.js';
import beyondPod from './beyond-pod/index.js';
import castro from './castro/index.js';
import clementine from './clementine/index.js';
import deezer from './deezer/index.js';
import downcast from './downcast/index.js';
import googlePodcasts from './google-podcasts/index.js';
import gpodder from './gpodder/index.js';
import iCatcher from './i-catcher/index.js';
import instacast from './instacast/index.js';
import overcast from './overcast/index.js';
import playerFm from './player-fm/index.js';
import pocketCasts from './pocket-casts/index.js';
import podcastAddict from './podcast-addict/index.js';
import podcastRepublic from './podcast-republic/index.js';
import podcat from './podcat/index.js';
import podscout from './podscout/index.js';
import procast from './procast/index.js';
import rss from './rss/index.js';
import rssRadio from './rss-radio/index.js';
import soundcloud from './soundcloud/index.js';
import spotify from './spotify/index.js';
import stitcher from './stitcher/index.js';
import youtube from './youtube/index.js';
import castbox from './castbox/index.js';
import { PodcatcherClientId } from './types.js';

export const CLIENTS = {
  'antenna-pod': antennaPod,
  'apple-podcasts': applePodcasts,
  'beyond-pod': beyondPod,
  castro: castro,
  clementine: clementine,
  deezer: deezer,
  downcast: downcast,
  'google-podcasts': googlePodcasts,
  gpodder: gpodder,
  'i-catcher': iCatcher,
  instacast: instacast,
  overcast: overcast,
  'player-fm': playerFm,
  'pocket-casts': pocketCasts,
  'podcast-addict': podcastAddict,
  'podcast-republic': podcastRepublic,
  podcat: podcat,
  podscout: podscout,
  procast: procast,
  rss: rss,
  'rss-radio': rssRadio,
  soundcloud: soundcloud,
  spotify: spotify,
  stitcher: stitcher,
  youtube: youtube,
  castbox: castbox
};

const match = <T>(matchers: null | T | T[], search: T): boolean => {
  if (!matchers) {
    return true;
  }

  if (Array.isArray(matchers)) {
    return matchers.includes(search);
  }

  return matchers === search;
};

const search =
  ({
    id = null,
    platform = null,
    title = null,
    type = null
  }: {
    id?: string | string[] | null;
    platform?: PodcastPlatform | PodcastPlatform[] | null;
    title?: string | null;
    type?: PodcastClientType | PodcastClientType[] | null;
  } = {}) =>
  (client: PodcastClient & { id: string }) => {
    const platformMatch = match(platform, client.platform);
    const typeMatch = match(type, client.type);
    const titleMatch = title
      ? (client.title || '').toUpperCase().includes(title.toUpperCase())
      : true;
    const idMatch = match(id, client.id);

    return idMatch && platformMatch && titleMatch && typeMatch;
  };

export default ({
  id,
  platform,
  title,
  type
}: {
  id?: PodcatcherClientId;
  platform?: PodcastPlatform | PodcastPlatform[];
  title?: string;
  type?: PodcastClientType | PodcastClientType[];
} = {}) =>
  Object.keys(CLIENTS)
    .reduce(
      (result: (PodcastClient & { id: string })[], id) => [
        ...result,
        ...CLIENTS[id].map((item) => ({ id, ...item }))
      ],
      []
    )
    .filter(search({ id, type, title, platform }));
