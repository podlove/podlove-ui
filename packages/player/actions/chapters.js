import { createAction } from './action-creator'

import {
  SET_CHAPTERS_LIST,
  NEXT_CHAPTER,
  PREVIOUS_CHAPTER,
  SET_CHAPTER,
  UPDATE_CHAPTER
} from './types'

export const setChapters = createAction(SET_CHAPTERS_LIST)
export const nextChapter = createAction(NEXT_CHAPTER)
export const previousChapter = createAction(PREVIOUS_CHAPTER)
export const setChapter = createAction(SET_CHAPTER)
export const updateChapter = createAction(UPDATE_CHAPTER)
