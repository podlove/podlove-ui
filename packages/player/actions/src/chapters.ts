import { createAction } from 'redux-actions';
import type { PodloveWebPlayerChapter } from '@podlove/types';

import {
  SET_CHAPTERS_LIST,
  NEXT_CHAPTER,
  PREVIOUS_CHAPTER,
  SET_CHAPTER,
  UPDATE_CHAPTER
} from './types';

export type setChaptersPayload = PodloveWebPlayerChapter[];
export type setChapterPayload = number;
export type updateChapterPayload = number;

export const setChapters = createAction<setChaptersPayload>(SET_CHAPTERS_LIST);
export const nextChapter = createAction<void>(NEXT_CHAPTER);
export const previousChapter = createAction<void>(PREVIOUS_CHAPTER);
export const setChapter = createAction<setChapterPayload>(SET_CHAPTER);
export const updateChapter = createAction<updateChapterPayload>(UPDATE_CHAPTER);
