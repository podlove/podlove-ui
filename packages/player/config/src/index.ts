import { compose, map, prop, propOr } from 'ramda';
import { toPlayerTime } from '@podlove/utils/time';
import { createObject } from '@podlove/utils/helper';
import type {
  PodloveWebPlayerAudio,
  PodloveWebPlayerChannel,
  PodloveWebPlayerChapter,
  PodloveWebPlayerConfig,
  PodloveWebPlayerFile,
  PodloveWebPlayerReference,
  PodloveWebPlayerResolvedConfig,
  PodloveRuntime,
  PodloveWebPlayerShare,
  PodloveWebPlayerSubscribeButton,
  PodloveWebPlayerTab,
  PodloveTheme,
  PodloveWebPlayerTranscript,
  PodloveWebPlayerPlaylistItem
} from '@podlove/types';

export const duration = compose<any[], string, number | null>(
  toPlayerTime,
  propOr('0', 'duration') as (input: PodloveWebPlayerConfig) => string
);

export const version = propOr<string | null>(null, 'version');

export const playtime = compose<any[], string, number | null>(
  toPlayerTime,
  propOr('0', 'playtime') as (input: PodloveWebPlayerConfig) => string
);

export const media = compose<any[], PodloveWebPlayerAudio[], PodloveWebPlayerAudio[]>(
  map(
    createObject({
      url: propOr(null, 'url'),
      size: propOr(0, 'size'),
      title: propOr(null, 'title'),
      mimeType: propOr(null, 'mimeType')
    }) as (input: any) => PodloveWebPlayerAudio
  ),
  propOr([], 'audio')
);

export const chapters = propOr([], 'chapters') as (input: PodloveWebPlayerResolvedConfig) => PodloveWebPlayerChapter[];

export const theme = (config: PodloveWebPlayerConfig): PodloveTheme => {
  const theme = propOr({}, 'theme', config) as (
    input: PodloveWebPlayerConfig
  ) => PodloveTheme;

  if (version(config) === 5) {
    return {
      tokens: propOr({}, 'tokens', theme),
      fonts: propOr({}, 'fonts', theme)
    };
  }

  const brand = propOr(null, 'main', theme) as string;

  if (brand) {
    return {
      tokens: {
        brand
      },
      fonts: {}
    };
  }

  return {
    tokens: {},
    fonts: {}
  };
};

export const reference = propOr({}, 'reference') as (
  input: PodloveWebPlayerResolvedConfig
) => PodloveWebPlayerReference;
export const transcripts = propOr([], 'transcripts') as (
  input: PodloveWebPlayerConfig
) => PodloveWebPlayerTranscript[];
export const shareReference = compose(
  propOr(null, 'share') as (input: PodloveWebPlayerReference) => string | null,
  reference
);

export const originReference = compose(
  propOr(null, 'origin') as (input: PodloveWebPlayerReference) => string | null,
  reference
);

export const episodeReference = (config: PodloveWebPlayerResolvedConfig): string | null => {
  const ref = reference(config);

  if (version(config) === 5) {
    return propOr(null, 'episode', ref);
  }

  return propOr(null, 'config', ref);
};

export const configReference = (config: PodloveWebPlayerResolvedConfig): string | null => {
  const ref = reference(config);

  if (version(config) === 5) {
    return propOr(null, 'config', ref);
  }

  return null;
};

export const validate = (config: PodloveWebPlayerResolvedConfig): boolean => {
  if (media(config).length === 0) {
    return false;
  }

  // Add further validations here ðŸ‘‡

  return true;
};

export const runtime = propOr({}, 'runtime') as (
  input: PodloveWebPlayerConfig
) => PodloveRuntime;

export const language = compose<any[], PodloveRuntime, string>(
  prop('language') as (input: PodloveRuntime) => string,
  runtime
);
export const platform = compose(prop('platform'), runtime);

export const playlist = propOr([], 'playlist') as (
  input: PodloveWebPlayerResolvedConfig
) => PodloveWebPlayerPlaylistItem[];

export const files = (config: PodloveWebPlayerResolvedConfig): PodloveWebPlayerFile[] => {
  const files: PodloveWebPlayerFile[] = propOr(media(config), 'files', config);
  return files
    .filter(({ url }) => !!url)
    .reduce(
      (result: PodloveWebPlayerFile[], item) => [
        ...result,
        ...(result.some(({ url }) => url === item.url) ? [] : [item])
      ],
      []
    );
};

export const activeTab = prop('activeTab') as (
  input: PodloveWebPlayerConfig
) => PodloveWebPlayerTab;

export const subscribeButton = prop('subscribe-button') as (
  input: PodloveWebPlayerConfig
) => PodloveWebPlayerSubscribeButton;

export const share = propOr({}, 'share') as (
  input: PodloveWebPlayerConfig
) => PodloveWebPlayerShare;

export const channels = compose<any[], PodloveWebPlayerShare, PodloveWebPlayerChannel[]>(
  propOr([], 'channels') as (input: PodloveWebPlayerShare) => PodloveWebPlayerChannel[],
  share
);

export const sharePlaytime = compose<any[], PodloveWebPlayerShare, boolean>(
  propOr(false, 'sharePlaytime') as (input: PodloveWebPlayerShare) => boolean,
  share
);
