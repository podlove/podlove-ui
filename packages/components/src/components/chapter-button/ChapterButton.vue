<template>
  <button :id="`chapter-button--${type}`" @click="clickHandler">
    <icon :type="type" :color="color" />
    <slot />
  </button>
</template>

<script>
import { color } from '../../defaults'
import Icon from '../icons/Icon'
import { nextChapter, previousChapter } from '@podlove/player-actions/chapters'

export default {
  components: {
    Icon
  },
  props: {
    type: {
      type: String,
      required: true,
      validator: (val) => ['next', 'previous'].includes(val)
    },
    color: {
      type: String,
      default: color
    }
  },
  emits: {
    next: null,
    previous: null
  },
  methods: {
    clickHandler() {
      switch (this.type) {
        case 'next':
          this.$emit('next', nextChapter())
          break
        case 'previous':
          this.$emit('previous', previousChapter())
          break
      }
    }
  }
}
</script>

<style lang="scss" scoped></style>
