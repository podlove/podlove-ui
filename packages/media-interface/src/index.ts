import { audio } from './audio.js';
import { video } from './video.js';
import { events } from './events.js';
import { actions } from './actions.js';
import { props } from './props.js';
import * as connect from './connect.js';

export { audio, video, events, props, actions, connect };
