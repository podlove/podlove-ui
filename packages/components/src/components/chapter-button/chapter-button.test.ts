import { mount } from '@vue/test-utils'
import { nextChapter, previousChapter } from '@podlove/player-actions/src/chapters'
import { describe, test, expect } from 'vitest';

import ChapterButton from './ChapterButton.vue'

describe('ChapterButton', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(ChapterButton, {
        props: {
          type: 'next'
        }
      })

      expect(wrapper.vm).toBeTruthy()
    })
  })

  describe('props', () => {
    const types = ['next', 'previous']

    types.forEach((type) => {
      test(`should render type '${type}' with prop color`, async () => {
        const wrapper = await mount(ChapterButton, {
          props: {
            color: '#fff',
            type
          }
        })

        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('events', () => {
    test('should emit NEXT_CHAPTER on next click', async () => {
      const wrapper = await mount(ChapterButton, {
        props: {
          type: 'next'
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted().next).toEqual([[nextChapter()]])
    })

    test('should emit PREVIOUS_CHAPTER on next click', async () => {
      const wrapper = await mount(ChapterButton, {
        props: {
          type: 'previous'
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted().previous).toEqual([[previousChapter()]])
    })
  })
})
