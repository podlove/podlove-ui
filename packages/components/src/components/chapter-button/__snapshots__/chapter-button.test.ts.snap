// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChapterButton > props > should render type 'next' with prop color 1`] = `
<button
  class="podlove-component-chapter-button"
  data-test="podlove-component-chapter-button--next"
>
  <svg
    aria-hidden="true"
    class="podlove-component-icon"
    data-v-258dccf6=""
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    <path
      d="M10 5.75263C10 5.18583 10.6045 4.82386 11.1042 5.09152L22.7659 11.3389C23.2937 11.6216 23.2937 12.3784 22.7659 12.6611L11.1042 18.9085C10.6045 19.1761 10 18.8142 10 18.2474V5.75263Z"
      fill="currentColor"
    />
    <path
      d="M22 5.75C22 5.33579 22.3358 5 22.75 5H23.25C23.6642 5 24 5.33579 24 5.75V18.25C24 18.6642 23.6642 19 23.25 19H22.75C22.3358 19 22 18.6642 22 18.25V5.75Z"
      fill="currentColor"
    />
    
  </svg>
  
  
</button>
`;

exports[`ChapterButton > props > should render type 'previous' with prop color 1`] = `
<button
  class="podlove-component-chapter-button"
  data-test="podlove-component-chapter-button--previous"
>
  <svg
    aria-hidden="true"
    class="podlove-component-icon"
    data-v-258dccf6=""
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    <path
      d="M14 5.75263C14 5.18583 13.3955 4.82386 12.8958 5.09152L1.23407 11.3389C0.706292 11.6216 0.706293 12.3784 1.23407 12.6611L12.8958 18.9085C13.3955 19.1761 14 18.8142 14 18.2474V5.75263Z"
      fill="currentColor"
    />
    <path
      d="M2 5.75C2 5.33579 1.66421 5 1.25 5H0.75C0.335787 5 0 5.33579 0 5.75V18.25C0 18.6642 0.335787 19 0.75 19H1.25C1.66421 19 2 18.6642 2 18.25V5.75Z"
      fill="currentColor"
    />
    
  </svg>
  
  
</button>
`;
